src/cvrp/ecvrp.py

```diff
@@ -80,58 +90,348 @@ def get_roads(self) -> tuple[tuple[int, ...], ...]:
+
+        This function will remove a point in a solution. This point comes from the road
+        which have been randomly selected from another solution.
+        The same process will be done for the other solution with a road which have been
+        randomly selected in this solution previously.
+        """
+        for road in solution:
+            if element in road:
+                road.remove(element)
+                break
+        return solution
+
+    def __tuple_to_list(self, tuple_element: tuple[tuple[int]]) -> list[list[int]]:
+        """Convert a tuple[tuple[int] into a list[list[int]]."""
+        return [list(x) for x in tuple_element]
+
+    def __merge_roads(self, solution: list[list[int]]) -> list[int]:
+        """
+        Merge roads to obtain a solution in a simple list.
+
+        The goal is to merge a list of road (which each is a list of int)
+        to obtain a solution which is a list of integer (ID point).
+        """
+        fusion_solution: list[int] = []
+        for index_road, road in enumerate(solution):
+            for index_point, point in enumerate(road):
+                if (index_point == 0) and (index_road == 0):
+                    fusion_solution.append(point)
+                elif index_point != 0:
+                    # the goal is to not have 0 in duplicate
+                    # which is all the time in index 0 for each road
+                    fusion_solution.append(point)
+        return fusion_solution
+
+    def __pull_off_chargers(self, road: list[int]) -> list[int]:
+        """
+        Remove all the chargers from a solution.
+
+        This function is helping the mutation process by removing all the chargers
+        of a road. It will retrun the road without it.
+        """
+        for point in road:
+            if self.__instance.is_charger(point):
+                road.remove(point)
+        return road
+
+    def __reversed_content(self, road: list[int], index1: int, index2: int) -> list[int]:
+        """Reverse the content of a road between the point 1 and 2 (included)."""
+        size = len(road)
+
+        if index1 < index2:
+            index_min = index1
+            index_max = index2
+        else:
+            index_min = index2
+            index_max = index1
+
+        begining = road[0:index_min]
+        middle = road[index_min:(index_max+1)]
+        middle.reverse()
+        ending = road[(index_max+1):size]
+
+        reversed_road = begining + middle + ending
+
+        return reversed_road
+
+    def __two_opt(self, road: list[int]) -> list[int]:
+        """
+        2-opt variance algorithm.
+
+        Algorithm which find the best mutation for this road by calculating
+        the distance for each mutation of road. The shortest distance will
+        determine the best muatation road.
+        """
+        best_distance = self._compute_road_fitness(road)
+        best_road = road
+
+        road.remove(0)  # remove the first depot
```

Comment:
The depot is not necessarily the index 0. It could also be the index 2 or 1025684.